src/channel-qvm.lisp

```diff
@@ -0,0 +1,417 @@
+   ;; Save a pointer to the originally provided memory
+   (slot-value qvm 'original-amplitude-pointer) (amplitudes qvm)))
+
+
+(defun requires-swapping-p (qvm)
+  "Does the noisy qvm QVM require swapping of internal pointers?"
+  (and (not (eq (amplitudes qvm) (original-amplitude-pointer qvm)))
+       #+sbcl (eq ':foreign (sb-introspect:allocation-information
+                             (original-amplitude-pointer qvm)))))
+
+
+(defmethod run :after ((qvm channel-qvm))
+  ;; Only copy if we really need to.
+  (when (requires-swapping-p qvm)
+    ;; Copy the correct amplitudes into place.
+    (copy-wavefunction (amplitudes qvm) (original-amplitude-pointer qvm))
+    ;; Get the pointer back in home position. We want to swap them,
+    ;; not overwrite, because we want the scratch memory to be intact.
+    (rotatef (amplitudes qvm) (%trial-amplitudes qvm))))
+
+
+
+(defmethod transition ((qvm channel-qvm) (instr quil:gate-application))
+  ;; If any noise rules are matched by theinstruction INSTR, apply the
+  ;; the kraus operators for that noise rule after applying the gate
+  ;; for this instruction.
+  (let* ((gate   (pull-teeth-to-get-a-gate instr))
+         (params (mapcar #'(lambda (p) (force-parameter p qvm))
+                         (quil:application-parameters instr)))
+         (instr-qubits (mapcar #'quil:qubit-index (quil:application-arguments instr)))
+         (noise-rules (noise-rules (noise-model qvm)))
+         (prepend-kraus-maps (match-rules noise-rules instr :before))
+         (append-kraus-maps (match-rules noise-rules instr :after)))
+    (when prepend-kraus-maps (apply-all-kraus-operators qvm instr prepend-kraus-maps))
+    (apply #'apply-gate gate (amplitudes qvm) (apply #'nat-tuple instr-qubits) params)
+    (when append-kraus-maps (apply-all-kraus-operators qvm instr append-kraus-maps))
+    (incf (pc qvm))
+    qvm))
+
+
+(defmethod transition :around ((qvm channel-qvm) (instr quil:measurement))
+  (let ((ret-qvm (call-next-method)))
+    (apply-classical-readout-noise ret-qvm instr)
+    ret-qvm))
+
+
+(defun rule-matches-instr-p (rule instr position)
+  "Check if rule is matched by instruction data and the position of the match request."
+  (let* ((noise-predicate (predicate (noise-predicate rule)))
+         (noise-position (noise-position (noise-predicate rule))))
+    (and (funcall noise-predicate instr) (string= position noise-position))))
+
+
+(defun match-rules (rules instr position)
+  "Return the operation elements for the first matching rule."
+  (loop :for rule in rules
+        :when (rule-matches-instr-p rule instr position)
+          :return (operation-elements rule)))
+
+
+(defmethod apply-all-kraus-operators ((qvm channel-qvm) (instr quil:gate-application) kraus-maps)
+  ;; Apply all the kraus operators to the state of the system.
+  (loop :for kraus-map :in kraus-maps
+        :do
+           (check-kraus-ops kraus-map)
```

Comment:
Would it suffice to move this `check-kraus-ops` into `make-noise-rule` ? And maybe change the `:accessor` for the `operation-elements` slot to a `:reader` ? In general I'm all for safety checks, but would be good if it can be safely moved out of this loop.